### ISSUE TITLE
How do unbound variables in rules work?

### ISSUE THREAD

**Issue by lvh** [tonsky/datascript#218]
Ex:

```clojure
[(f ?x ?y)
 [?z :a/x ?x]
 [?z :a/x ?y]]
```

I would expect that `?z` behaves like a gensym, where it has to be the same for the rule to match, but it doesn't collide with `?z` outside of the rule. However, that's just an expectation, I'm not sure if that's an expected property.

**Comment by tonsky** [comment]
No, that’s totally correct. It does work that way

**Comment by lvh** [comment]
Cool. In that case, I think I've found a bug. Here's a sample repro:

```clojure
(def prefix ["us" "il" "chicago" "get-gadget-attribute"])
(def db (->> (for [i (range 10)
                   attr-name [:has-gizmo :color]
                   :let [attr-val (case attr-name
                                    :has-gizmo (even? i)
                                    :color :red)
                         ;; the argument passed to the REST service in my db
                         args {:gadget-id i :attr-name attr-name}]]
               [(conj prefix args attr-name attr-val)
                (conj prefix args :gadget-id i)])
             (apply concat)))

(d/q '[:find ?id
       :where
       [?country ?state ?city "get-gadget-attribute" ?args :gadget-id ?id]
       [?country ?state ?city "get-gadget-attribute" ?args :has-gizmo true]]
     db)
;; => #{[4] [6] [8] [0] [2]} as expected

(d/q '[:find ?id
       :where
       [?country ?state ?city "get-gadget-attribute" ?args :gadget-id ?id]
       [?country ?state ?city "get-gadget-attribute" ?args :has-gizmo false]]
     db)
;; => #{[7] [9] [3] [5] [1]} as expected

(d/q '[:find ?id
       :in $ %
       :where
       (gadget-attr ?id :has-gizmo false)]
     db
     '[[(gadget-attr ?id ?attr ?val)
        [?country ?state ?city "get-gadget-attribute" ?args :gadget-id ?id]
        [?country ?state ?city "get-gadget-attribute" ?args ?attr ?val]]])
;; #{[4] [7] [6] [9] [3] [8] [0] [5] [2] [1]} ???

(d/q '[:find ?id
       :in $ %
       :where
       (gadget-attr ?id :has-gizmo true)]
     db
     '[[(gadget-attr ?id ?attr ?val)
        [?country ?state ?city "get-gadget-attribute" ?args :gadget-id ?id]
        [?country ?state ?city "get-gadget-attribute" ?args ?attr ?val]]])
;; #{[4] [6] [8] [0] [2]} as expected
```

Would you like me to file a separate issue? I think the issue is with falsy values; something is checking for boolean truth and it really wants to check with`nil?`/`some?`.

**Comment by lvh** [comment]
There's some extra noise in there that isn't necessary, so I narrowed it down to the minimal repro:

```clojure

(def db [[1 :x true] [2 :x false]])

(d/q '[:find ?id :where [?id :x true]] db)
(d/q '[:find ?id :where [?id :x false]] db)

(def rules '[[(prop ?id ?attr ?val) [?id ?attr ?val]]])

(d/q '[:find ?id :in $ % :where (prop ?id :x true)] db rules)
(d/q '[:find ?id :in $ % :where (prop ?id :x false)] db rules)
```

**Comment by tonsky** [comment]
Ok thanks will take a look

**Comment by tonsky** [comment]
Should be fixed in 0.16.1

**Comment by lvh** [comment]
Whoa, awesome. Thanks!
